### DISCUSSION TITLE
0077 XLS-77d Deep Freeze

### DISCUSSION THREAD

**Discussion by shawnxie999** [XRPLF/XRPL-Standards#220] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Thanks for the proposal!

Considering the title: may I propose that we change "blackholing" to "Permafrosting"? 😬

**Comment by tequdev** [comment]
Are there any usecases issuer sets only tfSetFreeze after this amendments enabled?

**Reply by shawnxie999** [reply]
In most scenarios, setting only tfSetFreeze is sufficient to meet compliance requirements and prevent transferring funds. But, there might be some rare scenarios or compliance needs where a deep-freeze is needed to ensure a blacklisted account can no longer receive or send funds.

**Reply by ckeshava** [reply]
Prior to this proposal, a regularly-frozen trust line can still recieve the (the frozen) IOU. However, the sanctioned counter-party cannot sell it. So, the sanctioned account could keep accumulating the IOUs and eventually send it back to the issuing address. Is this an accurate description of the workflow? Are there any other implications ?

**Comment by sappenin** [comment]
@shawnxie999 Is this one pretty baked? If so, want to open a PR for this one? (consider something like [this comment](https://github.com/XRPLF/XRPL-Standards/discussions/162#discussioncomment-10452843) if you want).

**Comment by shawnxie999** [comment]
Per the new [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), it is my opinion that we have reached a "well-refined standard."

As such, I propose that we move this discussion to a file (via https://github.com/XRPLF/XRPL-Standards/pull/260) and work on final changes using additional PRs, for better change-tracking.

Please comment here if you would like to object to moving this spec/discussion forward in the process into a DRAFT spec.

(Note that per the Contributing guidelines, moving a spec into the [DRAFT](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md#drafts) state does not mean any kind of endorsement, nor does it mean that this specification will become adopted. It is solely meant as a mechanism to enable better change tracking using PRs.)

**Comment by ckeshava** [comment]
1. `GlobalFreeze` applies to all the tokens issued by an address, whereas deep-freeze can be tailored for individual trust lines. 

`GlobalFreeze` suspends all Offers from selling the IOU under consideration. However, Offers which want to *buy* the IOU could still successfully complete. This is also a difference in behavior compared to `DeepFreeze`. Is this comparison accurate ?

2. If an individual address enables deep-freeze on its trust line with a financial institution, what are the implications on the relevant Offers? The regular-freeze feature does not impact Offers in this circumstance. Please see third paragraph from the end in this section: https://xrpl.org/docs/concepts/tokens/fungible-tokens/freezes#individual-freeze

3. Although the implementation has accounted for `Check` feature, the spec does not make a mention of `Check`s. This proposal modifies the behavior of `CheckCreate` and `CheckCash` transactions, is that right?

**Reply by ckeshava** [reply]
Regarding point-2, it appears that DeepFreeze differs from RegularFreeze whilst handling Offers. This is based on this implementation file: https://github.com/XRPLF/rippled/pull/5187/files#diff-aaa8dff0ed4bb7864a9d2aa7410ae8bde37b4ca91c7594413ab5691c0964e8e5

**Reply by shawnxie999** [reply]
1. Have you tested this scenario where offers can still buy an IOU despite being globally frozen? I don't believe this is possible because when an offer tries to _buy_ an IOU, there will always involve someone trying to _sell_, which should always fail.
2. It means the individual address can no longer transfer IOU with other holders. However, in the real world a holder should never be able to freeze the issuer.
3. Yes, it should prohibit cashing a check if the IOU is deep frozen. Spec needs to be updated

**Reply by vvysokikh1** [reply]
1. As far as current implementation goes, when global freeze is enacted, it is not possible at all to create new offers involving a globally frozen asset. So technically it's not possible to cross such existing offers.

**Reply by ckeshava** [reply]
1. No, I have not tested it. I'm basing my understanding off the documentation. I understand your point 👍 
2. Yes, I understand that. If an individual address deep-freezes a trust-line with an issuer, then all the outstanding offers for that specific IOU will become unfunded. This behavior is different with the regular-freeze, I wanted to clarify this aspect.
3. Alright, so only `CheckCash` transaction could throw an error with regards to a deep-frozen trust line. `CheckCreate` and `CheckCancel` do not throw any `tecFROZEN` errors. Is that correct ?

**Reply by shawnxie999** [reply]
2. Would be good to clarify with @vvysokikh1. I forgot what the behavior is when an individual address regular-freezes the issuer (doc may not necessarily be correct)
3. Yes, that's the decision for now. Only `CheckCash` is important to prevent transfer of funds.

**Comment by ximinez** [comment]
https://github.com/XRPLF/rippled/pull/5187 has been merged, so this can be merged as well.

**Comment by sappenin** [comment]
Closing this discussion since the formal XLS has been merged via #260 (spec is [here](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0077-deep-freeze)).
